readthedocs/search/tests/test_utils.py

```diff
@@ -0,0 +1,72 @@
+import pytest
+from django.core.urlresolvers import reverse
+
+from readthedocs.builds.constants import LATEST, STABLE
+from readthedocs.projects.models import HTMLFile, Project
+from readthedocs.search import utils
+from readthedocs.search.tests.utils import get_search_query_from_project_file
+
+
+@pytest.mark.django_db
+@pytest.mark.search
+class TestSearchUtils:
+
+    @classmethod
+    def setup_class(cls):
+        # This reverse needs to be inside the ``setup_class`` method because from
+        # the Corporate site we don't define this URL if ``-ext`` module is not
+        # installed
+        cls.url = reverse('doc_search')
+
+    def has_results(self, api_client, project_slug, version_slug):
+        query = get_search_query_from_project_file(
+            project_slug=project_slug,
+        )
+        search_params = {
+            'project': project_slug,
+            'version': version_slug,
+            'q': query
+        }
+        resp = api_client.get(self.url, search_params)
+        assert resp.status_code == 200
+
+        data = resp.data['results']
+        return len(data) > 0
+
+    def test_remove_only_one_project_index(self, api_client, all_projects):
+        project = 'kuma'
+
+        assert self.has_results(api_client, project, LATEST)
+        assert self.has_results(api_client, project, STABLE)
+
+        utils.remove_indexed_files(
+            HTMLFile,
+            project_slug=project,
+        )
+
+        assert not self.has_results(api_client, project, LATEST)
```

Comment:
So.. Looks like the deletion from indexes in ES happens async, so when checking here, it returns `True` instead of `False`. I was able to fix this locally with a `sleep(1)` before making a request in `has_results`.

Are we ok with the solution? Or is there any other way to test this?

Reply:
I'd say that it's OK for now with a comment in the code like this one.

Also, maybe 1s is too much if it will be called in more than one test. Maybe 0.5s still works.

Reply:
Hrm... This ended up being 3s finally, which seems to be too much and seeing that you changed to 1, 1.5, 2 and 3 it also seems that this may be unstable and the test will start failing randomly at some point.

What about just using mock and checking that the correct ES function was called? In the end, that's what we want to test, right?

Reply:
> In the end, that's what we want to test, right?

Nope, we want to test the ES response.

Reply:
Why would be want to test ES response? Testing ES response would be a test of ES code, not ours. If we already told ES to remove the index, that should be enough for us (we trust on ES :smile:). We just need to test that we are effectively telling ES to remove that index.

Reply:
I'm not trying to test ES, I'm trying to test the side effect of calling our function.

Reply:
I think we should just check the call we're making against ES -- we should be able to tell if it isn't filtered there, no?

Reply:
I'd prefer if we check ES, so we don't have surprises while upgrading our deps or ES itself, since the lib could produce an invalid query for the version of ES we have. Also, now with circleci, looks like 1 second is enough.

Reply:
> since the lib could produce an invalid query for the version of ES we have

If the lib we use produces an invalid query, it's not our fault either.

In any case, we could check that:

- we are calling the function (from the underlying lib) we should with the exact arguments
- check the HTTP request is _exactly_ what we expect and includes the filter

If the underlying version of ES changes and doesn't support the filter anymore, the request will fail.

Reply:
> check the HTTP request is exactly what we expect and includes the filter

We need to check that the request is what ES expect, we don't know the expected request. We should test the side effect (the deletion) as we do the same with everything else, we don't mock the ES requests anywhere else.